### DISCUSSION TITLE
Configuration

### DISCUSSION THREAD

**Discussion by Rongmario** [CleanroomMC/Cleanroom#65] [Draft]
### Summary

To provide a concise but extensive configuration API for devs, and for the players to edit config with ease whether in-game or in their filesystems.

### Goals

- Establish a coherent configuration API with zero-confusion. From constructing configurations programmatically, along with how it is translated into a GUI screen to how the actual configuration files outputted.
- Abstracted logic to allow developers to provide their own GUI screens to edit their own configs without having to worry about parsing configurations themselves.
- Allow configuration to be modified, viewed asynchronously.
- Utilize a file watcher to watch over the configuration directory.
- Give developers access to other mods' configurations with ease.
- Provide API to register ways of parsing configuration values into specific objects. And it should be Cleanroom's job to provide default parsers for Minecraft objects.
- Incorporate [ConfigAnytime's](https://github.com/CleanroomMC/ConfigAnytime) functionality natively. Configuration API shouldn't be late-loading.
- The configuration folder should be neater, files should match mod names, or at least in a nested folder if the mod wanted multiple configuration files that offers more explicitly categorized entries.

### Non Goals

- This is not to provide a convention for configuration formats (JSON/HOCON/TOML etc), that may come in a different CFP.
- We are not restricting developers to only be able to use this API for configuration, that is out of the scope and rather obtuse.

### Motivation

- Forge configurations had always been a "so close yet so far" situation with great annotation processing, yet it fails in so many places.
    - No public way of adding additional types to be parsed, only String, arrays and primitives by default.
    - ConfigManager loads in too late, and only works after asm data had been gathered. ConfigAnytime was made to circumvent this.
    - Relies heavily on annotations, whilst being under-documented; the other programmatic way isn't documented whatsoever.
    - Quirks in the system: classloading issues.
- We do not want to see the configuration folder become a wild-west like in imaginary version environments running on the Fabric loader.
- Fluent configuration standards are nice.

### Description

The typical configuration class looks like this for Forge mods:

```java
@Config(modid = "modid")
public class ConfigClass {
    
    public static boolean configBooleanProperty = true;
    public static int configIntProperty = 42;
    
    public static final SubCategory SUB_CATEGORY = new SubCategory();
    
    public static class SubCategory {
        
        public boolean innerProperty = false; // Must be non-static, referenced via the member field in the outer class
        
    }
    
    // This has to be here for the configuration to be referenced early
    // The mod must depend on ConfigAnytime, which is yet another dependency for such a simple task
    // This static block also has to be explicitly placed here
    static {
        ConfigAnytime.register(ConfigClass.class);
    }
    
}
```

and the outputted `modid.cfg` (the whitespace at the bottom isn't added in on purpose):

```ini
# Configuration file

general {
    B:configBooleanProperty=true
    I:configIntProperty=42

    sub_category {
        B:innerProperty=false
    }

}


```

I believe the current `@Config` annotation does serve well, without any extraneous getters/setters, you are able to set config properties, however, it isn't very reactive and you will have to refresh the `ConfigManager` of any changes.

```diff
-@Config(modid = "modid")
+@Config // mod id should be inferred
```

First of all, the mod id should be inferred; we should try to eliminate places to state the mod id other than the mod's main class and/or manifest. The value in those constant areas should propagate properly.

```diff
-public static boolean configBooleanProperty = true;
+public static final Observable<Boolean> configBooleanProperty = Observable.of(true);
```
(note: syntax isn't final!)
While more verbose, the field types should be converted to something that is [*observable*](https://en.wikipedia.org/wiki/Observer_pattern). Not only does this allow devs to react to configuration changes on an entry-to-entry basis, it also allows for file watchers to change in-memory configuration entries singularly.

For file watchers, since Java's APIs do not offer any type of callback. We will have to poll on a different thread to watch for configuration changes and react to these `Observables` on our own. This may be a drawback and may have to be an opted-in/opted-out service for each configuration class.

```diff
-public static class SubCategory {
-        
-        public boolean innerProperty = false; // Must be non-static, referenced via the member field in the outer class
-       
-}
+ ? Suggestions are welcome ?
```

Currently, inner classes seem fine, but it may depend on how complex the implementation ends up.

```diff
-ConfigAnytime.register(ConfigClass.class);
```

`ConfigAnytime` will be absorbed and replaced. It will have no place anymore where our configuration system can be loaded at any point in time. We will have a mock mod container so existing mods that depends on it do not break.

```java
/**
 * Abstracts the types of properties away. Higher level logic must prevent invalid data types.
 */
interface ITypeAdapter
{
    /**
     * Assigns the default value to the property
     * @param property the property whose default value will be assigned
     * @param value the default value
     */
    void setDefaultValue(Property property, Object value);

    /**
     * Sets the properties value.
     * @param property the property whose value will be set
     * @param value the set value
     */
    void setValue(Property property, Object value);

    /**
     * Retrieves the properties value
     * @param prop the property whose value will be retrieved
     * @return the properties value
     */
    Object getValue(Property prop);

    Type getType();

    boolean isArrayAdapter();
}
```

Forge currently offers `ITypeAdapter` but it is package-private, and never meant for devs to touch. But we will change that, and offer something more abundant. First, it definitely will have to be generic, for any type. The "adapter" should instead cover both serializing and deserializing. The "generic" way also means arrays won't have to be treated differently.

```java
interface TypeAdapter<T> {

    T deserialize(String key, String value); throws IllegalTypeDeserializationException
    
    String serialize(String key, T value); throws IllegalTypeSerializationException

}
```

More description describing more of the API in the coming days!

### Dependencies

- This may depend on the mod loading revamp CFP that will be made in the coming days.

### Guidelines

- [X] I have and will continue to adhere to the guidelines

**Comment by Ecdcaeb** [comment]
### `@Config`
Modders could select their file types here. 

#### Enum
(https://github.com/Ecdcaeb/Cleanroom/blob/config_filewapper/src/main/java/net/minecraftforge/common/config/Config.java#L54)

#### Registry
`java.lang.String` here. And get the `Coder` from a registry. Actually, enum is a static registry I throught.

### FileType
Do not breaking change the existing Configuration loading system, but remove `load` and `save`, and use `polymorphism` to support multiple file types.

#### JSON
JSON is not suitable as a configuration file, and we should not consider support for JSON.
- have not comment.
- not friendly for player
- not forced formated

There are lots of file types that are better than json, such as `toml` and `cfg`.

**Comment by Rongmario** [comment]
Though with the way you have done it, only `static final` enum values are accepted (since its part of an annotation). We would want it to be extensible, and why not JSON?

**Comment by ZZZank** [comment]
JSON is actually a good file format for devs, since it's usually structured to be similar to how corresponding Java class is structured. Designing via JSON will be easier then designing via Java.
I once came up with an idea on the Config system, and its data structure can be represented by JSON below: 
```json
{
    "percent": {
        "$comment": "single line comment",
        "$defaults": { "$min": 0, "$max": 100, "$value": 20 },
        "$value": 20
    },
    "enable": {
        "$comment": [
            "multi-line?"
        ],
        "$defaults": { "value": true },
        "$value": false
    },
    "someSubCategory": {
        "sub":{
            "$defaults": { "$all": ["NO", "YES"], "$value": "NO" },
            "$value": "NO"
        }
    }
}
```
## property
So for each entry in a config, sub-categories included, there should be:
- a name, e.g. `precent`, that can be used for indexing  
- default setups represented by `$defaults`, and  
- the actual value represented by `$value`.  
Java class may look like this: 
```java
interface IConfigProp<T> {
    T get(); //get value
    String getName(); //get name
    UnmodifiableList<String> getComments();
    IConfigDefaults<T> getDefaults(); //get default setup
    T_ <T_> getAs(Class<T_> type);
}
```
T can be category, so sub-category is supported. 
## comments
I recommend using List<String> or at least String[] for comments, since representing multi-line comments(`new ArrayList<>()`), single line comment(`Collections.singletonList()`) and no comment(`Collections.emptyList()`) can all be easy. 
## defaults
`$defaults` here, should be another Java class that is responsible for checking if the value read from file is valid, and providing default value when there is no valid value. 
```java
interface IConfigDefaults<T> extends Predicate<T> {
    boolean test(T value);
    T getDefaultValue();
}
```
All "bounds" related data, like `$min` `$max` ranges for int, or `$all` for enum, should be hold by IConfigDefaults.

**Comment by Rongmario** [comment]
I personally am against JSON because of the poor readability format with the stringent syntax, a minor typo can be hard to diagnose especially with Gson's error messages when parsing.

The interfaces looks good and aren't verbose.

As I've written in the OP, the goal isn't to propose a configuration language/format/denotation (JSON or what not). That can come in a separate proposal (I know that most people in the discord have expressed likeability and interest towards TOML and HOCON).

**Reply by ZZZank** [reply]
I'm using JSON here just because it's handy for illuatrating my idea on the data structure of config internals, actual config should indeed be format-independent. And glad to hear that `The interfaces looks good` :)

**Reply by Rongmario** [reply]
Good stuff ^^

**Comment by Ecdcaeb** [comment]
I want to be able to trace the owner of a Configuration, like:

```java
public class ModConfiguration extends Configuration {
    public final String modId;
    public final Class<?> clazz;

    public ModConfiguration(String modidIn, Class<?> clazzIn, File file){
        super(fileWrapper);
        modId = modidIn;
        clazz = clazzIn;
    }
}
```

When creating a class-based config for an FMLMod：
```java
- cfg = new Configuration(file);
+ cfg = new ModConfiguration(modid, cls, file);
```

This is just a simple example, and other methods can also be used to record. However, the two existing Maps cannot be traced back.

```java
    private static Map<String, Configuration> CONFIGS = Maps.newHashMap();
    private static Map<String, Set<Class<?>>> MOD_CONFIG_CLASSES = Maps.newHashMap();
```

Creating an additional Map can also solve the problem, but I think creating and using subclasses is more in line with human logic.

**Comment by Ecdcaeb** [comment]
~ConfigAnytime aims to enable configurations to be registered at any time. How can this be achieved when applied to decoders? Here is a perverted example: our decoder itself also relies on configuration files! ConfigAnytime requires that it can be used at any time, even if fml mods are not in the classpath. We may need a plugin or service that is earlier than Coremod to accomplish this, which can satisfy the use of coremod and mods, but still cannot satisfy the perverted example above. What I want to ask is: Can Anytime in ConfigAnytime have an upper limit?~

the upper limit is `the static block of LoadPlugin`, we may use Service to complete the registration before this.

**Reply by kappa-maintainer** [reply]
The path info needed by ConfigManager was set in FMLTweaker, butForge doesn't even begin coremod discovery when FMLTweaker was run.
Example: Fugue sync config at the static block of LoadPlugin, which is the earliest point any coremod could reach.
Please read source before commenting.

**Reply by Ecdcaeb** [reply]
This is a design confirmation.

**Comment by Ecdcaeb** [comment]
Currently, the class-based configuration method is inconvenient to migrate. If I want to modify its format, I still need to keep the old fields for compatibility. This leads to many problems:
1. It doesn't look good
2. If the player does not actively modify it, the old configuration fields will always exist
3. Config version cannot be used in this situation. In another API it works fine.

Therefore, we need: `ConfigFixUpper`

**Comment by Bronitt** [comment]
Recently, I’ve been intrigued by the idea of creating a configuration system that would cover all possible developer needs. You may have already seen one of my attempts at implementing this on the Discord server (TOMLConfig), but at the time I ran into limitations of the TOML format.

We usually think of a Map as
```
map: [
    key1: value1,
    key2: value2
]
```
And so on, but in TOML there are two main implementations of a Map:
```TOML
inline_map = { key1 = value1, key2 = value2 }

[standard_map]
key1 = value
key2 = value
```
As you can see, the multi-line map closely resembles ***category***, and distinguishing between them in a config file just by looking at it will be quite difficult.

That’s exactly why I’m currently leaning toward creating custom format, and I think the standard `.cfg` file from Forge is a good starting point. However, I want to make a few changes to the format itself and completely rewrite the parser. One of my ideas is to add a generated configuration header that will contain the modId, modName, and modVersion of the mod to which this configuration file belongs. This will spare modpack developers from having to specify the “source” of the configuration. Additionally, the version of the format itself can be embedded in the header so that outdated configurations can be tracked (if the format itself changes) and outdated parts can be updated “on the fly,” avoiding the need to include an outdated parser in every new version. This also means that existing .cfg files will be automatically converted to the new format the first time the game is launched (at the very beginning, the parser will check the format version, and if it encounters a problem it cannot handle, the entire file or a specific field will be marked as corrupted, and after the game loads, a notification about the problem will appear in the menu. In this case, a default value will be substituted for the in-game values)

### Why did I decide to bring this up here in the first place?
I’m looking for suggestions from the community and the developers themselves regarding the new format. My goal is to create a user-friendly tool for every developer based on familiar tools, without making significant changes to the APIs they’re used to, and while maintaining backward compatibility with existing configurations. I don't plan on working specifically on a parser for the file itself anytime soon; rather, I'll be building an architecture into which any parser for any file can be integrated, so there's no need to worry about radical changes. So, if you have any ideas or suggestions for improving the new format, I’d love to hear your thoughts. 

You can find information about the project and a brief list of tasks in this repository: https://github.com/Quantum-Spiral-Team/ReConfiguration

*P.s. However, I'm still not sure that a custom format would offer more than, say, HOCON or JSON5, so I'd like to ask you once again to share your thoughts on this matter.*

**Reply by Bronitt** [reply]
As for storing and processing values, I suggest using an approach similar to the one used in `fastutil`: we have a general abstract class that looks like this
```java
public abstract class ConfigProperty {
    private final String name;
    private final List<String> comments;
    private final Class<?> type; 

    protected ConfigProperty(String name, List<String> comments, Class<?> type) {
        this.name = name;
        this.comments = comments;
        this.type = type;
    }

    public abstract boolean isValid();

    public String getName() { return this.name; }
    public List<String> getComments() { return this.comments; }
    public Class<?> getType() { return this.type; }
}
```
as well as separate implementations for objects:
```java
public class ObjectConfigProperty<T> extends ConfigProperty {
    private volatile T value;
    private final @Nullable Predicate<String> validator;

    public ObjectConfigProperty(String name, List<String> comments, Class<?> type, T defaultValue, @Nullable Predicate<String> validator) {
        super(name, comments, type);
        this.value = defaultValue;
        this.validator = validator;

        if (!isValid()) {
            throw new IllegalArgumentException("Default value for property '" + name + "' is invalid");
        }
    }

    public T get() { return this.value; }

    public boolean set(T value) {
        if (validator == null || validator.test(PropertyUtils.serialize(value))) {
            this.value = value;
            return true;
        } else {
            LOGGER.error("Failed to set property '{}': invalid value '{}'", getName(), value);
            return false;
        }
    }

    @Override
    public boolean isValid() {
        return validator == null || validator.test(PropertyUtils.serialize(this.value));
    }
}
```
and for primitives:
```java
public class IntConfigProperty extends ConfigProperty {
    private volatile int value;
    private final @Nullable LongPredicate validator; // LongPredicate for byte, short, int and long; DoublePredicate for float and double

    public IntConfigProperty(String name, List<String> comments, int defaultValue, @Nullable LongPredicate validator) {
        super(name, comments, int.class);
        this.value = defaultValue;
        this.validator = validator;

        if (!isValid()) {
            throw new IllegalArgumentException("Default value for property '" + name + "' is invalid");
        }
    }

    public int get() { return this.value; }

    public boolean set(int value) {
        if (validator == null || validator.test(value)) {
            this.value = value;
            return true;
        } else {
            LOGGER.error("Failed to set property '{}': invalid value '{}'", getName(), value);
            return false;
        }
    }

    @Override
    public boolean isValid() {
        return validator == null || validator.test(this.value);
    }
}
``` 

This way, we avoid converting values such as `int` <-> `Integer`, `short` <-> `Short`, etc.

I haven't fully worked out the logic behind `PropertyUtils` yet, but, roughly speaking, it's a class responsible for serializing and deserializing non-primitive objects so they can be written to the configuration. Among other things, you can use it to create your own serializer and deserializer implementations that are unique to your specific mod (to avoid conflicts when multiple mods try to write their own serializers for the same class). All of this is still in the design phase, so I can’t offer you a finished implementation just yet.

**Reply by Rongmario** [reply]
`ConfigProperty` seemed fine other than the obvious need for a way to declare if it had been modified or not (so the whole config can be marked dirty).

As for the format of config, it isn't part of this proposal, but if this groundwork is laid, any format can be used really.

**Reply by Bronitt** [reply]
> other than the obvious need for a way to declare if it had been modified or not (so the whole config can be marked dirty).

Yes, I agree that this is necessary, but I haven't decided on the best way to implement it yet. At the moment, I'm leaning toward an approach where all modified fields are collected into a single set and updated simultaneously (for example, when ConfigChangedEvent.OnConfigChangedEvent is triggered).

> As for the format of config, it isn't part of this proposal, but if this groundwork is laid, any format can be used really.

Yes, that's exactly what I'm working on right now. I want to split the configuration architecture into a "parser" and a "binder", so that we can replace or add a new "parser" at any time without affecting the "binder".

In any case, I'm currently working on "binder", trying to completely rewrite the vanilla parser (to make it faster and more flexible)

**Reply by gardenevery** [reply]
Thoughts on configuration format

file name: modid-{common, client, server}.toml

```
[group1]
    # comment
    # Type: Integer
    # Default: 1, Min: 0, Max: 2147483647
    config1 = 32

    # comment
    # Type: Boolean
    # Default: false
    config2 = true

[group2]
    # comment
    # Type: Enum (EnumDifficulty)
    # Default: NORMAL, Allowed: PEACEFUL, EASY, NORMAL, HARD
    config1 = "NORMAL"
```

Among which:
```
# Type: Integer
# Default: 1, Min: 0, Max: 2147483647

# Type: Boolean
# Default: false

# Type: Enum (EnumDifficulty)
# Default: NORMAL, Allowed: PEACEFUL, EASY, NORMAL, HARD
```
Should be auto-generated

**Reply by Bronitt** [reply]
@gardenevery 
This approach is similar to modern versions of the game. My goal isn’t to port new configurations, but to implement my own variation.
For example, personally, I don’t see the point in dividing configurations into `server`/`client`/`common`. After all, in version 1.12, developers almost never use this approach (although it is certainly possible). In my opinion, the best solution would be to give developers the option to choose whether to apply the `@SideOnly()` annotation to a configuration class, a category class, or a field. 
Furthermore, as I’ve already mentioned, maps are implemented in TOML (in the context of configurations) in a rather cumbersome way, and I’m sure many developers would gladly take advantage of the opportunity to use maps directly, rather than having to parse strings. Therefore, I personally oppose TOML and insist on preserving Forge’s original format while improving it.

Among other things, one of my goals is to maintain backward compatibility for other mods. Developers who want to use TOML usually install third-party libraries for this purpose (of which there are quite a few), so I don’t see the point in such radical changes (switching from `.cfg` to `.toml`)

**Reply by Rongmario** [reply]
Devs don't use this approach currently because it doesn't exist btw. It is very useful to have sided configurations and also helps compartmentalize which config option is applied on the client, server or both.

**Reply by Bronitt** [reply]
All right, then I don't think it will be a problem to add that method to `@Config`, like 
```Java
/**
 * {@code Side.CLIENT} - {@code {name}-client.cfg}
 * {@code Side.SERVER} - {@code {name}-server.cfg}
 */
Side[] sides() default {Side.CLIENT, Side.SERVER};
```
or just `@SideOnly()` on class/field

**Reply by Rongmario** [reply]
> All right, then I don't think it will be a problem to add that method to `@Config`, like
> ...

What about per-world configurations? Which is what Forge defaults server configurations to be in meme versions.

**Reply by Bronitt** [reply]
> What about per-world configurations? Which is what Forge defaults server configurations to be in meme versions.

Yeah, I've already thought about that. I'm planning to have at least three entry points: `preInit`, `<clinit>`, and `per-world`

**Reply by Rongmario** [reply]
`preInit` and `<clinit>`? what?

**Reply by Bronitt** [reply]
> `preInit` and `<clinit>`? what?

What I meant was `preInit` as the standard entry point in Foge configurations and `<clinit>` as in ConfigAnytime (but without having to manually call the method in `static {}`)

**Comment by kappa-maintainer** [comment]
I want to add multi-line entry support to string and string list now, just for a project I am working on

**Reply by Rongmario** [reply]
Entry?

**Reply by Bronitt** [reply]
Most likely, he means multi-line string support like
```
value = """
    First line
    Second line
    Third line
    etc.
"""
```

**Reply by Rongmario** [reply]
This is possible when writing
```
key = value1, value2, value3, \
          value4, value5, value6, \
          value7
```

as stated as of Java's `Properties` javadocs.

**Reply by kappa-maintainer** [reply]
Not in `Configuration`
I want something like
```
S:LIST <
    entry:value,value,value,
		  value,value,value,
		  value,value,value,
	entry2:value
>
```
And the first string will be read as `entry:value,value,value,value,value,value,value,value,value,`

**Reply by Rongmario** [reply]
afaik Forge uses it in the backend, did you try?

**Reply by kappa-maintainer** [reply]
Only splash and some other place use java properties, Config just loads `\` literal

**Reply by Rongmario** [reply]
Remind me not to read `Configuration::load` again, horrendous.

**Comment by Bronitt** [comment]
Regarding `@Config` automatically filling in the `modId` - how could that be implemented? To be honest, I can't think of anything other than iterating through the packages of all mods from `Loader.instance().getModList()`. Are there alternative ways to get the `modId` from a class with this annotation at runtime? The main issue, really, is when loading happens before `preInit`, like in ConfigAnytime.

**Comment by Ecdcaeb** [comment]
Make `@Config.AutoSync`, equals to 
```java
@SubscribeEvent
public static void onConfigChanged(final ConfigChangedEvent.OnConfigChangedEvent event) {
       if (event.getModID().equals(Reference.MOD_ID)) {
              ConfigManager.sync(Reference.MOD_ID, Config.Type.INSTANCE);
       }
}
```

**Reply by Rongmario** [reply]
I would prefer this to be the default

**Reply by Ecdcaeb** [reply]
Many mods have contains codes like this..... May conflicts
